compiler/vm/vmdef.nim

```diff
@@ -90,27 +91,224 @@ type
+  MemoryHandle* = object
+    # XXX: without this dummy pointer, memory gets corrupted in ways
+    # that lead to random illegal storage accesses. The corruption very likely
+    # still happens but it seems to be more hidden. It's likely that some part
+    # of the VM writes past their the allocated memory range
+    _: pointer
+    p: ptr Atom
+    l: int
+  VmMemoryRange* = openArray[byte] # TODO: rename to `VmMemoryRegion`
+
+  TypedHandle* = tuple[h: MemoryHandle, typ: PVmType]
+
+  VmSeq* = object
+    length*: int
+    data*: MemoryHandle
+
+  VmString* = distinct VmSeq
+
+  VmSet* = object
+    # XXX maybe use `ptr IntSet`? Can't use IntSet with automatic storage duration since
+    # sizeof(IntSet) is also huge (smaller than set[uint16] however)
+
+    # XXX should `setElems` be ordered? Would speed up contains checks but require sorting on modification
+    # XXX `setElem` could also be run-length encoded for speed and memory efficiency...
+
+    # XXX In the end, VmSets should be embedded into their surrounding storage, without any heap allocations
+    elems* : seq[uint16]
+
+  Atom* {.union.} = object
+    intVal*: BiggestInt ## akInt
+    floatVal*: BiggestFloat ## akFloat
+    ptrVal*: ptr Atom ## akPtr
+    setVal*: VmSet ## akSet
+    strVal*: VmString ## akString
+    seqVal*: VmSeq ## akSeq
+    refVal*: GlobalRef ## akRef
+    callableVal*: VmFunctionObject ## akCallable
+
+    nodeVal*: PNode ## akPNode
+
+  # TODO: needs a better name
+  VmObjectAtom* = object ## A container that stores a value. Single atom
+                         ## values are stored in-place while complex values
+                         ## are seperately allocated 
+    case kind*: AtomKind
+    of akInt..pred(akObject):
+      atom: array[32, byte]
+    of akObject, akArray:
+      obj*: MemoryHandle
+    
+    typ*: PVmType
+
+  
+  # XXX: VmObjectAtom doesn't store pseudo atoms in-place, due to their dynamic size.
+  # However, `TypedVmObject`s are only meant to be used for globals and constants. For these, 
+  # the pseudo-atom flat byte size is both constant and known at initialization time. Thus, 
+  # we could allocate the raw atom or object (pseudo atom) in the same memory location. We'd need a custom
+  # allocator for `TypedVmValue` then, but globals and constants probably need one anyway...
+  TypedVmValue* = VmObjectAtom
+
+  TFullReg* = object
     case kind*: TRegisterKind
-    of rkNone: nil
+    of rkNone: discard
     of rkInt: intVal*: BiggestInt
     of rkFloat: floatVal*: BiggestFloat
-    of rkNode: node*: PNode
-    of rkRegisterAddr: regAddr*: ptr TFullReg
-    of rkNodeAddr: nodeAddr*: ptr PNode
-
+    of rkAddress:
+      addrVal*: MemoryHandle
+      addrTyp*: PVmType
+      isValid*: bool ## Indicates if `addrVal` points to a valid location
+    of rkHandle, rkValue:
+      handle*: TypedHandle
+      global*: Option[GlobalRef] ## If valid, the global of which the handle references a location
```

Comment:
Since it's an index, you should allocate a known-invalid index (if it's an uint, just allocate the highest index) instead of using `Option` since it doubles your storage cost due to the `ok: bool` in it.

Reply:
The known-invalid index is zero (meaning nil). I temporarily use an `Option` here as an aid during debugging (could have also used a distinct with an accessor function). Once everything is stable it's going to just be `GlobalRef` (renamed it to `HeapSlotHandle`)